File: backend/MetalurgicaWeb/Metalurgica.API/Services/ProductoService.cs
using Metalurgica.API.DTOs.Producto;
using Metalurgica.API.Models;
using Metalurgica.API.Repositories.Interfaces;
using Metalurgica.API.Services.Interfaces;

namespace Metalurgica.API.Services
{
    public class ProductoService : IProductoService
    {
        private readonly IProductoRepository _repository;

        public ProductoService(IProductoRepository repository)
        {
            _repository = repository;
        }

        public async Task<IEnumerable<ProductoResponseDto>> GetAllAsync()
        {
            var productos = await _repository.GetAllAsync();
            return productos.Select(MapToResponseDto);
        }

        public async Task<ProductoResponseDto?> GetByIdAsync(int id)
        {
            var producto = await _repository.GetByIdAsync(id);
            return producto is null ? null : MapToResponseDto(producto);
        }

        public async Task<ProductoResponseDto?> GetByCodigoAsync(string codigo)
        {
            var producto = await _repository.GetByCodigoAsync(codigo);
            return producto is null ? null : MapToResponseDto(producto);
        }

        public async Task<ProductoResponseDto> CreateAsync(ProductoRequestDto dto)
        {
            if (await _repository.ExistsCodigoAsync(dto.Codigo))
                throw new InvalidOperationException($"Ya existe un producto con el código '{dto.Codigo}'.");

            var producto = new Producto
            {
                Nombre = dto.Nombre,
                Descripcion = dto.Descripcion,
                Precio = dto.Precio,
                StockActual = dto.StockActual,
                StockMinimo = dto.StockMinimo,
                Activo = dto.Activo,
                Codigo = dto.Codigo
            };

            await _repository.AddAsync(producto);
            await _repository.SaveChangesAsync();

            return MapToResponseDto(producto);
        }

        public async Task<bool> UpdateAsync(int id, ProductoRequestDto dto)
        {
            var producto = await _repository.GetByIdAsync(id);
            if (producto is null) return false;

            producto.Nombre = dto.Nombre;
            producto.Descripcion = dto.Descripcion;
            producto.Precio = dto.Precio;
            producto.StockActual = dto.StockActual;
            producto.StockMinimo = dto.StockMinimo;
            producto.Activo = dto.Activo;
            producto.Codigo = dto.Codigo;

            _repository.Update(producto);
            await _repository.SaveChangesAsync();
            return true;
        }

        public async Task<bool> DeleteAsync(int id)
        {
            var producto = await _repository.GetByIdAsync(id);
            if (producto is null) return false;

            _repository.Delete(producto);
            await _repository.SaveChangesAsync();
            return true;
        }

        private static ProductoResponseDto MapToResponseDto(Producto producto)
        {
            return new ProductoResponseDto
            {
                ProductoId = producto.ProductoId,
                Nombre = producto.Nombre,
                Descripcion = producto.Descripcion,
                Precio = producto.Precio,
                StockActual = producto.StockActual,
                StockMinimo = producto.StockMinimo,
                Activo = producto.Activo,
                Codigo = producto.Codigo
            };
        }
    }
}

File: frontend/src/services/producto.service.ts
import { apiClient } from '@/api/client';
import type { ProductoRequestDto, ProductoResponseDto } from '@/types/producto.types';

const BASE_PATH = '/api/Producto';

export const productoService = {
  getAll: async (): Promise<ProductoResponseDto[]> => {
    const response = await apiClient.get<ProductoResponseDto[]>(BASE_PATH);
    return response.data;
  },

  getById: async (id: number): Promise<ProductoResponseDto> => {
    const response = await apiClient.get<ProductoResponseDto>(`${BASE_PATH}/${id}`);
    return response.data;
  },

  create: async (dto: ProductoRequestDto): Promise<ProductoResponseDto> => {
    const response = await apiClient.post<ProductoResponseDto>(BASE_PATH, dto);
    return response.data;
  },

  update: async (id: number, dto: ProductoRequestDto): Promise<void> => {
    await apiClient.put(`${BASE_PATH}/${id}`, dto);
  },

  delete: async (id: number): Promise<void> => {
    await apiClient.delete(`${BASE_PATH}/${id}`);
  },
};